server/src/services/documentSymbol/onDocumentSymbol.ts

```diff
@@ -71,50 +71,91 @@ export function onDocumentSymbol(serverState: ServerState) {
-          new FallbackFunctionDefinition(document, builder),
-          new LibraryDefinition(document, builder),
-          new ModifierDefinition(document, builder),
-          new ReceiveFunctionDefinition(document, builder),
-          new UserDefinedValueTypeDefinition(document, builder),
-          new YulFunctionDefinition(document, builder),
-          new UnnamedFunctionDefinition(document, builder),
+        const finders: SymbolFinder[] = [
+          new StructDefinition(),
+          new StructMember(),
+          new InterfaceDefinition(),
+          new FunctionDefinition(),
+          new ContractDefinition(),
+          new EventDefinition(),
+          new StateVariableDefinition(),
+          new VariableDeclarationStatement(),
+          new ConstantDefinition(),
+          new ConstructorDefinition(),
+          new EnumDefinition(),
+          new ErrorDefinition(),
+          new FallbackFunctionDefinition(),
+          new LibraryDefinition(),
+          new ModifierDefinition(),
+          new ReceiveFunctionDefinition(),
+          new UserDefinedValueTypeDefinition(),
+          new YulFunctionDefinition(),
+          new UnnamedFunctionDefinition(),
         ];
 
-        const indexedVisitors = _.keyBy(visitors, "ruleKind");
-
         const cursor = parseOutput.createTreeCursor();
-        const ruleKinds = visitors.map((v) => v.ruleKind);
 
         // Useful to keep this here for development
-        // const kursor: Cursor = parseTree.cursor.clone();
+        // const kursor = cursor.clone();
         // do {
         //   console.log(
-        //     `${"  ".repeat(kursor.pathRuleNodes.length)}${kursor.node.kind}(${
-        //       ["R", "T"][kursor.node.type]
-        //     }): ${kursor.node?.text ?? ""}`
+        //     `${"  ".repeat(kursor.ancestors().length)}${kursor.node().kind}(${
+        //       // ["R", "T"][kursor.node().type]
+        //       kursor.node().type
+        //     })(${kursor.label}): ${(kursor.node() as TokenNode)?.text || ""}`
         //   );
         // } while (kursor.goToNext());
 
-        span = transaction.startChild({ op: "walk-generate-symbols" });
-        while (cursor.goToNextRuleWithKinds(ruleKinds)) {
-          const node = cursor.node() as RuleNode;
+        span = transaction.startChild({ op: "run-query" });
+
+        // Execute a single call with all the queries
+        const queries = finders.map((v) => Query.parse(v.query));
+        const results = cursor.query(queries);
+
+        span.finish();
+
+        span = transaction.startChild({ op: "build-symbols" });
 
-          const visitor: SymbolVisitor = indexedVisitors[node.kind];
-          visitor.onRuleNode(cursor);
+        // Transform the query results into symbols
+        let result;
+        const symbols = [];
+
+        while ((result = results.next())) {
+          for (const finder of finders) {
+            const symbol = finder.onResult(result);
```

Comment:
FYI, each query result has a `queryNumber` which is the index to the query that produced this result in the array above. You can use that if you want to call the correct `finders[i]` directly instead of calling `onResult()` on each of them, and skipping on all except the right one.

Reply:
Thanks for addressing!